server/src/test/java/org/opensearch/gateway/ShardBatchCacheTests.java

```diff
@@ -0,0 +1,231 @@
+        VALID
+    }
+
+    public void setupShardBatchCache(String batchId, int numberOfShards) {
+        Map<ShardId, ShardAttributes> shardAttributesMap = new HashMap<>();
+        fillShards(shardAttributesMap, numberOfShards);
+        this.shardCache = new AsyncShardBatchFetch.ShardBatchCache<>(
+            logger,
+            "batch_shards_started",
+            shardAttributesMap,
+            "BatchID=[" + batchId + "]",
+            GatewayStartedShard.class,
+            new GatewayStartedShard(null, false, null, null),
+            GatewayStartedShard::isEmpty,
+            new ShardBatchResponseFactory<>(true)
+        );
+    }
+
+    public void testClearShardCache() {
+        setupShardBatchCache(BATCH_ID, NUMBER_OF_SHARDS_DEFAULT);
+        ShardId shard = shardsInBatch.iterator().next();
+        this.shardCache.initData(node1);
+        this.shardCache.markAsFetching(List.of(node1.getId()), 1);
+        this.shardCache.putData(node1, new NodeGatewayStartedShardsBatch(node1, getPrimaryResponse(shardsInBatch, ResponseType.EMPTY)));
+        assertTrue(
+            this.shardCache.getCacheData(DiscoveryNodes.builder().add(node1).build(), null)
+                .get(node1)
+                .getNodeGatewayStartedShardsBatch()
+                .containsKey(shard)
+        );
+        this.shardCache.deleteShard(shard);
+        assertFalse(
+            this.shardCache.getCacheData(DiscoveryNodes.builder().add(node1).build(), null)
+                .get(node1)
+                .getNodeGatewayStartedShardsBatch()
+                .containsKey(shard)
+        );
+    }
+
+    public void testGetCacheData() {
+        setupShardBatchCache(BATCH_ID, NUMBER_OF_SHARDS_DEFAULT);
+        ShardId shard = shardsInBatch.iterator().next();
+        this.shardCache.initData(node1);
+        this.shardCache.initData(node2);
+        this.shardCache.markAsFetching(List.of(node1.getId(), node2.getId()), 1);
+        this.shardCache.putData(node1, new NodeGatewayStartedShardsBatch(node1, getPrimaryResponse(shardsInBatch, ResponseType.EMPTY)));
+        assertTrue(
+            this.shardCache.getCacheData(DiscoveryNodes.builder().add(node1).build(), null)
+                .get(node1)
+                .getNodeGatewayStartedShardsBatch()
+                .containsKey(shard)
+        );
+        assertTrue(
+            this.shardCache.getCacheData(DiscoveryNodes.builder().add(node2).build(), null)
+                .get(node2)
+                .getNodeGatewayStartedShardsBatch()
+                .isEmpty()
+        );
+    }
+
+    public void testInitCacheData() {
+        setupShardBatchCache(BATCH_ID, NUMBER_OF_SHARDS_DEFAULT);
+        this.shardCache.initData(node1);
+        this.shardCache.initData(node2);
+        assertEquals(2, shardCache.getCache().size());
+
+        // test getData without fetch
+        assertTrue(shardCache.getData(node1).getNodeGatewayStartedShardsBatch().isEmpty());
+    }
+
+    public void testPutData() {
```

Comment:
can we also assert on fetchingDone as both the nodes responded?

Reply:
Ack, added assertion in `testPutData` method.